tls/cfsslscan_handshake.go

```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
 	if err != nil {
 		return
 	}
+	if CipherSuites[serverHello.cipherSuite].EllipticCurve {
+		// Prime the connection, if necessary, for key
+		// exchange messages by reading off the certificate
+		// message and, if necessary, the OCSP stapling
+		// message
+		var msg interface{}
+		msg, err = c.readHandshake()
+		if err != nil {
+			return
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
+			msg, err = c.readHandshake()
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
+			if !ok {
+				c.sendAlert(alertUnexpectedMessage)
+				return
+			}
+		}
+
+		var skx *serverKeyExchangeMsg
+		skx, err = c.exchangeKeys()
+		if skx.raw[0] != typeServerKeyExchange {
+			err = errors.New("tls: received something other than the expected ServerKeyExchange.")
+			return
+		}
+		if err != nil {
+			return
+		}
+		if len(skx.key) < 4 {
+			err = errServerKeyExchange
+			return
+		}
+		curveType = uint16(skx.key[0])
+		// If we have a named curve, report which one it is.
+		if curveType == 3 {
+			curveID = CurveID(skx.key[1])<<8 | CurveID(skx.key[2])
+		}
+	}
 	cipherID, version = serverHello.cipherSuite, serverHello.vers
 	return
-
 }
 
 // sayHello is the backend to SayHello that returns a full serverHelloMsg for processing.
```

Comment:
Sometimes (commonly, but not uniformly), this function (specifically, the call to `c.readHandshake()`) returns the error `remote error: handshake failure` which is passed all the way out to the output. I can't discern any pattern in the requests that cause this error, other than that all the requests that trigger it have (a) only an EC cipher suite requested and (b) all possible curves. But not all such requests trigger the error.